File: solutions/csharp/interest-is-interesting/1/InterestIsInteresting.cs
using System;

static class SavingsAccount
{
    public static float InterestRate(decimal balance)
    {
        if (balance < 0m)
            return 3.213f;
        else if (balance < 1000m)
            return 0.5f;
        else if (balance < 5000m)
            return 1.621f;
        else
            return 2.475f;
    }

    public static decimal Interest(decimal balance)
    {
        return (decimal) InterestRate(balance) / 100m * balance;
    }

    public static decimal AnnualBalanceUpdate(decimal balance)
    {
        return balance + Interest(balance);
    }

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        decimal currentBalance = balance;
        int years = 0;
        while (currentBalance < targetBalance)
        {
            currentBalance = AnnualBalanceUpdate(currentBalance);
            years++;
        }
        return years;
    }
}

File: solutions/fsharp/interest-is-interesting/1/InterestIsInteresting.fs
module InterestIsInteresting

let interestRate (balance: decimal): single =
    match balance with
        | b when b < 0.0m -> 3.213f
        | b when b < 1000.0m -> 0.5f
        | b when b < 5000.0m -> 1.621f
        | _ -> 2.475f

let interest (balance: decimal): decimal =
   0.01m * decimal (interestRate balance) * balance

let annualBalanceUpdate(balance: decimal): decimal =
   balance + interest balance

let amountToDonate(balance: decimal) (taxFreePercentage: float): int =
    if balance > 0.0m then
        int (0.02m * decimal (taxFreePercentage) * balance)
    else
        0